tensorflow/python/ops/image_ops_impl.py

```diff
@@ -1086,8 +1086,9 @@ def convert_image_dtype(image, dtype, saturate=False, name=None):
         return math_ops.multiply(cast, scale, name=name)
       else:
         # Converting from float: first scale, then cast
-        scale = dtype.max + 0.5  # avoid rounding problems in the cast
+        scale = dtype.max
         scaled = math_ops.multiply(image, scale)
+        scaled = scaled + 0.5  # avoid rounding problems in the cast
```

Comment:
See my comment, I am not sure this is correct.

Reply:
Think about the case, when you converting a `uint8` image from RGB to grayscale. You first convert the image to `float32`, then do the conversion, you may get the value like `0.9/255`. when converting back to `uint8`, it supposed to be rounded to `1`. I guess this conversion bug will cause mistakes in many other image operations, like resize.

Reply:
I suppose whether 0.9/255 should be 0 or 1 depends on whether you want your int domain to consist of equal-sized buckets. The current behavior implements a floor. I believe this is the correct behavior, since it maps 0..1 into 256 equal-sized buckets. 

If you use round behavior as you suggest, [0..0.5) is mapped to 0, then you have 254 buckets of size 1, and finally another bucket of size 0.5. 

I believe floor is the better solution here. However, I recognize this could be considered  a matter of taste. We cannot change behavior though, so we should leave this alone.

Reply:
But the old code doesn't have equal-sized buckets either.
For example, when converting from `244.9/255` in `float32` to `uint8`, first multiply with `255 + 0.5` and get `245.4`, then floor to `245`. However, it supposed to be `244`

Reply:
You are right. 

0.5 is a very crude epsilon. We could reduce it to a very small number. The point of adding it in the first place was to avoid 1/255 + 3/255 = 3.9999/255 ==> 3/255. But it will never be perfect. I'm happy to reduce it, or if you have a fundamentally better approach, I'd be excited about that too.